Add new preferences persistence API, and save editor preferences in user meta

talldan opened the pull request:
Related - https://github.com/WordPress/gutenberg/issues/31965
Also related - https://github.com/WordPress/gutenberg/pull/39672#discussion_r832967226

Fixes #15105
Fixes #34405

## What?
This PR explores moving away from the current `@wordpress/data` way of persisting store data to local storage.

It implements a new API in the preferences package for configuring how the store state is persisted.

It adds a new package that persists preferences to the WordPress database, as part of the user's meta.

And then finally this is all configured using an inline script.

## Why?
A good reason is that this will finally allow https://github.com/WordPress/gutenberg/issues/15105 to be closed, which is a long standing problem for users.

There are also some technical advantages that this will unlock:
- deprecating the old persistence system in `@wordpress/data`
- deprecating `registerStore`in `@wordpress/data` in favor of `register`

## How?
Adds a new `setPersistenceLayer` action to the preferences package. A higher order reducer is then used to keep state in sync with the persistence layer in the preferences store.

There's also a new `preferences-persistence` package (based on https://github.com/WordPress/gutenberg/pull/19177).  This saves data to user meta via the REST API primarily. It also saves data to local storage as a back in case the user goes offline or a request is interrupted.

Finally, I've moved all the migration code over from the old package to this new package, any old preferences should be retained.

## Testing Instructions
1. Open up two different browsers side by side.
2. Login to your test environment in both browsers using the same login and edit a post, site or widgets
3. Change some settings in one browser (i.e. top toolbar)
4. Reload the other browser and you should see the change reflected.

talldan committed:
Migrate block editor insert usage to preferences store

Update tests

Make updateInsertUsage a proper action that can be unit tested

Fix reusable block tests

talldan committed:
Implement basic API

talldan committed:
Add new local storage persistence package

talldan committed:
Only call persistenceLayer.set if there is a persistence layer

talldan committed:
Wire up dependencies

talldan committed:
Fix exporting create as default and get not being async

talldan committed:
Remove old persistence interface and configure new persistence layer in post editor

talldan committed:
Use correct user id and storage key

talldan commented:
The more I'm working on this, the more I'm thinking that the ideal setup will be two packages:
- `wordpress/preferences` - a general purpose key/value data store that persists to local storage out of the box, but the persistence layer is configurable. Has no WordPress-specific code.
- `wordpress/wp-preferences` - WordPress specific utils for `wordpress/preferences`. Contains a persistence layer for the WordPress database, all the code for the WordPress data migrations, and exports react components that use preferences. It could possibly have its own php script for preloading preferences that are in the database.

This would entail moving the react components that are already in `wordpress/preferences` to this new package, but that should be ok if they're deprecated for two plugin versions and don't land in core. I think it'd be good to keep the main preferences package only as a data store (reducing it's only proper dependency to `@wordpress/data`).

talldan committed:
Move local storage persistence layer into the preferences package

talldan committed:
Try a wp-preferences package

talldan committed:
Use database persistence

talldan committed:
Only set preference values with particular actions

talldan committed:
Remove component for now

youknowriad commented on packages/edit-post/src/utils/configure-preferences.js:
```diff
@@ -0,0 +1,14 @@
+import { createDatabasePersistenceLayer } from '@wordpress/wp-preferences';
+
+export default async function configurePreferences( defaults ) {
+	const databasePersistenceLayer = createDatabasePersistenceLayer();
+	await dispatch( preferencesStore ).setPersistenceLayer(
+		databasePersistenceLayer
+	);
```
Should this be in edit-post or just an `add_inline_script` for the `preferences` package instead? (kind of like we configure apiFetch which is generic to be wp-specific)

youknowriad commented on packages/wp-preferences/src/database-persistence-layer/index.js:
```diff
@@ -0,0 +1,41 @@
+ */
+import apiFetch from '@wordpress/api-fetch';
+
+export default function createDatabasePersistenceLayer() {
```
I kind of don't like `wp-preferences` as a package name. Instead, could `database-persistence-layer` be a package instead?

it could also be `@wordpress/preferences-api-adapter` and be configurable if needed (API path...) and all the gluing happens as an inline script in the server to make it WP specific.

It can even be something the raw preferences package exposes (the configurable adapter) since `api-fetch` is not WordPress specific by default.

Is that too complex?

talldan replied:
A better name would be good. I also want to think about somewhere to put all the migration code that's in the data package at the moment - https://github.com/WordPress/gutenberg/blob/trunk/packages/data/src/plugins/persistence/index.js#L588, I'm not sure yet whether it's right for this to live alongside the database persistence layer.

I was also considering moving the components out of the preferences package to somewhere else, but that's not entirely necessary.

Using an inline script would also be good. I think it might be best for preferences to do its own preloading and not rely on apiFetch, and this would be a good way to do it.

talldan committed:
Make native initialize function async

talldan committed:
Revert changes to preferences package.json

talldan committed:
Try using an inline script to configure preferences

talldan commented on packages/edit-post/src/index.js:
```diff
@@ -10,6 +10,7 @@ import { render, unmountComponentAtNode } from '@wordpress/element';
 import { dispatch, select } from '@wordpress/data';
 import { addFilter } from '@wordpress/hooks';
 import { store as preferencesStore } from '@wordpress/preferences';
+import '@wordpress/database-persistence-layer';
```
Adding this was the only way I found to make the package available as `wp.databasePersistenceLayer` for the inline script.

There's probably another way, it would be great to hear if there is.

talldan commented on packages/database-persistence-layer/src/index.js:
```diff
@@ -0,0 +1,17 @@
+export { create };
+
+export function __experimentalConfigureDatabasePersistenceLayer( options ) {
+	const persistenceLayer = create( { options } );
+	dispatch( preferencesStore ).setPersistenceLayer( persistenceLayer );
```
Initially I wanted this code to be in the inline script, but I couldn't figure out how to make an inline script have multiple dependencies.

Exposing one function from the package seemed to be the easiest option, but that means adding `data` and `preferences` as dependencies to the package, which otherwise wouldn't have been necessary.

youknowriad replied:
I think in my mind it would have been something like that:

```php
wp_add_dependency_to_script( 'wp-preferences', 'wp-database-persistence-layer' );
wp_add_inline_script( 'wp-preferences', 'wp.data.dispatch( wp.preferences.store ).setPersistenceLayer( wp.databasePersistenceLayer.create() );', 'after' );
```

In the above I'm adding `wp-database-persistence-layer` as a dependency to `wp-preferences` but it can also be done at registration for `wp-preferences` in the script loader.

talldan commented:
@youknowriad This is now working using an inline script, but there might be a better way of doing it (see my two comments above).

Seems to work pretty well in usage though. We can also preload the preferences via the script, which might be a safer option compared to relying on the api fetch preloading.

talldan committed:
Use a filter to add the necessary script dependencies

talldan committed:
Remove import and package.json dependency from edit-post in favor of wiring up scripts in PHP

talldan committed:
Remove merging of state

talldan committed:
Try abort controller for cancelling multiple in flight requests

talldan committed:
Revert "Try abort controller for cancelling multiple in flight requests"

This reverts commit 06345b57253f80c3619be3e34c6c5b0e2d6aa0b9.

youknowriad commented on lib/compat/experimental/persisted-preferences.php:
```diff
@@ -0,0 +1,113 @@
+	$preload_data = get_user_meta( $user_id, $meta_key, true );
+
+	wp_add_inline_script(
+		'wp-database-persistence-layer',
```
I don't understand why you chose to add this to `wp-database-persistence-layer` instead of `wp-preferences` (which mean you don't have to add any dependencies to any other script than `wp-preferences`)

talldan replied:
Hmm, true. I'm still learning about the best way to do this. Thanks for the pointers.

talldan committed:
Try throttling API requests

talldan committed:
Simplify dependencies

youknowriad commented on packages/database-persistence-layer/README.md:
```diff
@@ -0,0 +1,3 @@
+# Database persistence layer
```
Thinking about the name here:

 - Should the name include "preferences" to clarify that it's about persisting preferences
 - Should the name include "api" instead of "database" (I think that's probably arguable)?

talldan replied:
> Should the name include "preferences" to clarify that it's about persisting preferences

I'd like to add data migration code to the package that's specific to preferences, so I think it makes sense to mention `preferences`.

A problem with `api` and `database` is that those names are about the implementation. If we wanted to change it in future we'd probably need a new package.

We could go with something like `preferences-persistence-layer` and in the README state that this is the persistence layer for WordPress preferences. Maybe it exports `createUserMetaPersistenceLayer`, and if in the future we want to persist preferences in a different way, a new function can be added and the old one can be deprecated.

youknowriad replied:
the problem with `preferences-persistence-layer` is that it a local storage is also a persistence layer. I'm fine keeping "database" or "server" or something like that.

gziolo replied:
> A problem with api and database is that those names are about the implementation. If we wanted to change it in future we'd probably need a new package.

It's hard to imagine that this implementation would change for WordPress in the future. Anyway, based on the keywords used in the code and your discussion I wanted to check what do you think about the following ideas:
- `@wordpress/user-preferences`
- `@wordpress/preferences-user-persistence`

talldan replied:
I quite like those two suggestions @gziolo. Do you have a preference @youknowriad?

youknowriad replied:
What about just `@wordpress/preferences-persistence` ?

noisysocks replied:
My 2¢ on naming:

Agree with Riad's thinking. There's two things to name.

The first thing is the abstraction we're introducing to `@wordpress/preferences`. You've chosen _persistence layer_. I think it might be clearer to call this something like _preferences backend_ or _preferences store_. Using the word _preferences_ makes it clear that `@wordpress/preferences` is what we're talking about.

The second thing is the type of the above abstraction that puts preferences into the database. You've chosen _database_ which I think is a little misleading because it really uses the REST API. I doubt that will ever change.

So I'd go with `rest-api-preferences-store` or something along those lines.

---

**BUT** all the names suggested so far are a bit weird when you consider that this new package also contains all of the migrations for legacy local storage preferences (`src/migrations/legacy-local-storage-data`). Maybe that stuff should move to to the package where the legacy calls to `localStorage` used to live? (`@wordpress/editor` I think?)

talldan replied:
> You've chosen database which I think is a little misleading because it really uses the REST API. I doubt that will ever change.

I'm not so sure about that, a switch to GraphQL, websockets, or webRTC could all make that name irrelevant. It would be a very big change for WordPress, but then using a realtime communication protocol is very likely to become a reality for collaborative editing. While it's unlikely everything would go through such an API, I think it's enough of a reason not to consider 'REST API' in the naming.

It's this kind of reason that most other packages are named for what they do rather than how they do it. There are exceptions, but most of those are very small utils. Because of that I would favour something like `preferences-persistence` or `preferences-user-persistence`.

I think I'll go with `preferences-persistence` as Riad suggested, just to keep it simple.

youknowriad reviewed (commented):
This is looking very promising for me

talldan committed:
Test that preferences are not persisted in tests due to reload

talldan commented:
I think the reason why a lot of the tests have been failing is due to the way the page reloads after toggling a preference here in the `createNewPost` util:
https://github.com/WordPress/gutenberg/blob/88689f06fed418a07b842b83ddbfd5153601f604/packages/e2e-test-utils/src/create-new-post.js#L46-L53

I think that's preventing the API request that would persist the change to preferences. After the page reloads it's like `toggleFeature` never happened.

An option could be to use `navigator.sendBeacon` for the API requests - https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon. This still sends a request even when the page unloads. It would mean missing out on APIFetch middleware.

Or maybe we just accept that this may happen and make the tests more robust. 🤔 

edit: Another option still could be to use local storage as well as a fallback layer. There would need to be a way to know whether local storage is more recent than the database.

talldan commented:
Still a lot of e2e tests failing. I think another reason for that might be that the tests clear `localStorage` which in `trunk` would reset preferences:
https://github.com/WordPress/gutenberg/blob/606788a48eec4130309402be68c6a87612b3fb9e/packages/e2e-tests/config/setup-test-framework.js#L78-L81

But here the preferences are persisted between test runs in the database.

talldan committed:
Reset all preferences in e2e test setup

talldan committed:
Try `keepalive`

talldan committed:
Oh, this is actually a debounce, not a throttle

talldan commented:
Still getting a lot of test failures.

There are also still some tests that do a `page.reload` and are they are racing and winning against the API request to user meta (specifically the one to disable the welcome guide that every test does). Because the preference isn't saved in time, when the editor reloads the welcome guide is unexpectedly open.

There are other test failures that are a bit more obscure. I'm having a hard time reproducing and/or understanding those.

It'd be good to get an idea of whether there are specifically bugs that are likely to happen for humans, or if these failures are only a result of the tests running so quickly.

talldan committed:
Try a localstorage fallback

talldan committed:
Avoid `null` value becoming preloaded state

talldan committed:
Improve naming and docs

talldan committed:
Revert "Test that preferences are not persisted in tests due to reload"

This reverts commit 3865610409ae2ce960ade361f1e75e213263da11.

talldan committed:
Use a stable reference when storage is empty

talldan committed:
Also reset preferences after each test

talldan committed:
Use beforeEach for clearing e2e test browser state

talldan committed:
Try avoiding beforeAll and afterAll in nonce test

talldan committed:
Revert "Use beforeEach for clearing e2e test browser state"

This reverts commit e51a00a08bdfba21fcc2d2598d428ef7740d28f2.

talldan committed:
Switch to using beforeEach and afterEach, as they also ensure correct order of operation

talldan committed:
Revert "Migrate block editor insert usage to preferences store"

This reverts commit da21c6793cc7d2beeb084c23260004131c8624bc.

talldan commented:
I think the e2e tests are mostly passing now. The tests did need some extra setup code to clear server stored preferences before and after test runs, so that was one of the main things that was missing.

Even then, solely using an async server API wasn't stable enough, there were still issues like page reloads in the tests racing and beating the preferences API request, which meant after the reload the editor state would be incorrect.

These would have been edge cases for human users, but to make things more robust I've made the preferences persist to user meta and also local storage as a backup. If for some reason the user meta preferences are outdated, the local storage will be used instead.

youknowriad commented:
> there were still issues like page reloads in the tests racing and beating the preferences API request, which meant after the reload the editor state would be incorrect.

I feel like I had some things like that in page reloads for other stuff as well. It might good to have our own dedicated "reload" helper where we could inject things safely and not update every test at some point. (or maybe you did that I didn't check :P )

> These would have been edge cases for human users, but to make things more robust I've made the preferences persist to user meta and also local storage as a backup. If for some reason the user meta preferences are outdated, the local storage will be used instead.

This is cool but makes me wonder if it's better to actually "preload" the preferences and make them "blocking" for the UI. Only show the UI once they're loaded. I'm thinking about folks loading their editor in a new browser and noticing a "jump" when the preferences load. Still not sure what's best though.

talldan commented:
> I feel like I had some things like that in page reloads for other stuff as well. It might good to have our own dedicated "reload" helper where we could inject things safely and not update every test at some point. (or maybe you did that I didn't check :P )

@youknowriad Do you remember what the other stuff was? I did think about this option. Waiting for network requests to complete before a reload might be a start.

> This is cool but makes me wonder if it's better to actually "preload" the preferences and make them "blocking" for the UI. Only show the UI once they're loaded. I'm thinking about folks loading their editor in a new browser and noticing a "jump" when the preferences load. Still not sure what's best though.

The good news is that they are preloaded.

The issue is more that when the page reloads, a request to save preferences can be in flight and not finish before the  page has reloaded. This is especially true because some tests are toggling a lot of preferences quickly (opening and closing panels and sidebars), and the server requests are debounced to avoid overwhelming the server, so the last timeout or request can be delayed or interrupted.

youknowriad commented:
> Do you remember what the other stuff was?

I don't know but for me, it was more about doing something after the "reload". For instance I see `await page.waitForSelector( '.edit-post-layout' );` is almost always called right after the reload function.

> The good news is that they are preloaded.
> The issue is more that when the page reloads, a request to save preferences can be in flight and not finish before the page has reloaded. This is especially true because some tests are toggling a lot of preferences quickly (opening and closing panels and sidebars), and the server requests are debounced to avoid overwhelming the server, so the last timeout or request can be delayed or interrupted.

👍

talldan committed:
Migrate basic preferences in from localstorage

talldan committed:
Port the migrate-individual-preference function

talldan committed:
Port the migrate-feature-preference function

talldan committed:
Copy convert-edit-post-panels function from persistence plugin

talldan committed:
Add moveInterfaceEnableItems function and improve folder structure

talldan committed:
Tidy up migration code

talldan committed:
Add a test for the whole migration and fix a bug the test found

talldan committed:
Improve comment

talldan committed:
Fix some more migration issues

talldan committed:
Remove old migration code

talldan commented:
This seems to be working well now. I've added some code to migrate existing local storage preferences to the new system.

To finish up I'll add docs, tests and changelog updates. Then I'll consider splitting the PR up into smaller PRs (it is quite big, though it's mostly tests at the moment).

There's also a need to consider the naming of the new package, so I'll put some more thought into that.

talldan committed:
Add some docs for setPersistenceLayer

talldan committed:
Update database persistence layer docs

talldan committed:
Add preferences package changelog entries

talldan committed:
Remove unused local storage persistence folder

talldan committed:
Remove unused local storage persistence layer

talldan committed:
Add unit tests to preferences store

talldan committed:
Add tests for persistence layer

talldan committed:
Handle equality when comparing timestamps

talldan committed:
Improve the preferences README further

talldan committed:
Improve docs

talldan committed:
Move php code into wordpress-6.1 folder

talldan commented:
This should be ready for review now. It is quite a big PR, though lots of it is tests and docs. The main thing that's still under discussion is the new package name, but the implementation will only change now based on review feedback.

If anyone is brave enough, we can review this all in one go and try to merge the one PR.

Alternatively I've broken out part of this into a separate PR - https://github.com/WordPress/gutenberg/pull/40246. That just contains the new API in the preferences package as the first part and should be easier to review.

talldan commented on packages/e2e-tests/config/setup-test-framework.js:
```diff
@@ -242,6 +243,7 @@ beforeAll( async () => {
 	enablePageDialogAccept();
 	observeConsoleLogging();
 	await simulateAdverseConditions();
+	await resetPreferences();
```
Playwright tests are passing, but I think this will probably needed to be added to the configuration for Playwright too.

Not sure of the best way. @kevin940726 any advice?

kevin940726 replied:
Yeah, we can simply port `resetPreferences` to Playwright using `requestUtils`. And if it's something that needs to be cleared for every test then put it in `test.ts`'s `requestUtils` fixture.

talldan replied:
I've added that in 8b472ca 👍

talldan committed:
Reset preferences when running playwright tests

TimothyBJacobs commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+ */
+function gutenberg_configure_persisted_preferences() {
+	$user_id = get_current_user_id();
+	if ( empty( $user_id ) ) {
```
We should always call `register_meta`, even if the user isn't logged in.

noisysocks replied:
Maybe put the `register_meta` stuff into a `init` hook and the inline script into a `admin_init` hook. There should always be a logged-in user when `admin_init` is invoked.

TimothyBJacobs replied:
I think that's a good idea. Though for clarity `admin_init` won't necessarily have a logged in user. It will fire for `admin-ajax.php` and `admin-post.php` which can be used by logged out users.

TimothyBJacobs commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+				'name'   => 'persisted_preferences',
+				'type'   => 'object',
+				'schema' => array(
+					'type'                 => 'object',
```
We should add a `context` of `edit`, otherwise they could be publicly exposed.

TimothyBJacobs commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+	wp_add_inline_script(
+		'wp-preferences',
+		sprintf(
+			'const serverData = %s;
```
Should this section of code be wrapped in something like an IIFE? I think this ends up globalizing all of these variables.

noisysocks replied:
Yeah typically we wrap with `( function() { } )()` and use ES5-style JavaScript so that there's some compatibility with ancient browsers. (Technically we don't support those browsers but I think it's nice to do anyway.)

talldan replied:
Yes, very true, thanks for mentioning it.

I guess it won't be compiled either, so I should make it ye olde JavaScript.

TimothyBJacobs commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+				'type'   => 'object',
+				'schema' => array(
+					'type'                 => 'object',
+					'properties'           => array(),
```
From my reading of the JS, it looks like `__timestamp` is a property that is always set. We should document it as a property then.

TimothyBJacobs commented on packages/database-persistence-layer/src/create/index.js:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	function set( newData ) {
+		const dataWithTimestamp = { ...newData, __timestamp: Date.now() };
```
Typically we transport dates in the REST API as `RFC3339`.

noisysocks replied:
`toISOString()` ought to do what @TimothyBJacobs suggests.

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+
+	global $wpdb;
+
+	$meta_key = $wpdb->get_blog_prefix() . 'persisted_preferences';
```
Why do we need to prefix? For multisites?

talldan replied:
Yep, there was some feedback about it on the old PR - https://github.com/WordPress/gutenberg/pull/19177#issuecomment-567601834.

I'll add an inline comment so that it's clear to readers.

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+			const localData = JSON.parse(
+				localStorage.getItem( localStorageRestoreKey )
+			);
+			const serverTimestamp = serverData?.__timestamp ?? 0;
```
I'm not certain we can use `??` in non-transpiled code.

talldan replied:
I've switched it to a `||`, which works better now we have to use Date.parse, since that returns a falsey `NaN`.

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+				preloadedData = __unstableConvertLegacyLocalStorageData( userId );
+			}
+
+			const { create } = wp.databasePersistenceLayer;
```
I'm not certain we can use object destructuring in non-transpiled code.

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,102 @@
+			$user_id
+		),
+		'after'
+	);
```
In my opinion this feels like too much code for an inline script. Inline scripts have limitations: they aren't transpiled and they aren't cached by the browser. Consider putting the bulk of this logic into a function that's in `@wordpress/database-persistence-layer` and having the inline script merely call that.

```js
wp.databasePersistence.initPersistenceLayer( serverData, userId );
```

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
````diff
@@ -0,0 +1,102 @@
+ * case 'wp-preferences':
+ *     array_push( $dependencies, 'wp-database-persistence-layer' );
+ *     break;
+ * ```
````
🙌 you are a good person, thank you.

talldan committed:
Always register user meta

noisysocks commented on packages/database-persistence-layer/src/create/create-async-debounce.js:
```diff
@@ -0,0 +1,55 @@
+ * @return {Function} A function that debounce whatever function is passed
+ *                    to it.
+ */
+export default function createAsyncDebounce( delayMS ) {
```
Let's add unit tests for this. (Alternatively, is there a npm package we can use which has equivalent functionality?)

noisysocks replied:
Maybe https://www.npmjs.com/package/debounce-async.

talldan replied:
Just tried that library so that I don't have to write tests, but it seems buggy for a leading edge debounce. It alternates between leading and trailing. I can see why—it never clears the promise after resolution of a leading edge debounce, so it falls through to trailing on every other call.

The package seems unmaintained, so probably no point making an upstream fix, but I made an issue - https://github.com/szchenghuang/debounce-async/issues/5

noisysocks replied:
I think I saw another package in npm if you wanted to try that.

talldan replied:
It was painful, but I managed to write some tests.

noisysocks commented on packages/database-persistence-layer/src/create/create-async-debounce.js:
```diff
@@ -0,0 +1,55 @@
+	let timeoutId;
+	let activePromise;
+
+	return async function debounce( func ) {
```
Why a higher order function instead of `asyncDebounce( func, wait )` like in lodash? Thinking that keeping the API similar (identical?) to lodash will make it more comfortable for developers to use.

talldan replied:
I've made it work like that 👍

noisysocks commented on packages/database-persistence-layer/src/create/index.js:
```diff
@@ -0,0 +1,105 @@
+			return cache;
+		}
+
+		const user = await apiFetch( {
```
This will throw if there's a network error, if the server has an unexpected error (`500`), or if the user was logged out (`400`). Is that right?

talldan replied:
I'll check this out.

noisysocks commented on packages/database-persistence-layer/src/create/index.js:
```diff
@@ -0,0 +1,105 @@
+		const localData = JSON.parse(
+			localStorage.getItem( localStorageRestoreKey )
+		);
+		const serverTimestamp = serverData?.__timestamp ?? 0;
```
Stupid nit but probably one `_` is fine.

noisysocks commented on packages/database-persistence-layer/src/create/index.js:
```diff
@@ -0,0 +1,105 @@
+		);
+
+		// The user meta endpoint seems susceptible to errors when consecutive
+		// requests are made in quick succession. Ensure there's a gap between
```
That really ought to be fixed on the server... maybe create a Trac ticket 😀

TimothyBJacobs replied:
Core is pretty full of APIs that are subject to race conditions. Without rearchitecting how metadata works, I'm not sure there would be a solution there that would allow for two requests to be processed at the same time and not lead to data loss or errors.

noisysocks commented on packages/database-persistence-layer/src/index.js:
```diff
@@ -0,0 +1,2 @@
+export { default as create } from './create';
+export { default as __unstableConvertLegacyLocalStorageData } from './migrations/legacy-local-storage-data';
```
Why move all of this?

I mentioned this already in another comment but it's a little odd that this package purports to be an implementation of a database persistence layer but then also includes all of this migration code.

Should it be in a new package?

Or could it be in the package where all of these old preferences used to live e.g. `@wordpress/editor`?

talldan replied:
I think the migration code has always been in `@wordpress/data` before this with the persistence plugin, but I don't think that was good because that package isn't concerned with the concrete structure of data. I think this is at least a step in the right direction, because this package contains lots of things specific to WordPress. There are also a few small changes to the data to make it work with the new persistence API.

I don't think `@wordpress/editor` is an option because it's unused by the widgets editors.

I did consider that each package that implements a preference could have migration code for that preference. That seems like the right ownership, but gluing all that together would be complicated.

I guess it could be an API like this:
```js
wp.preferences.registerMigration( oldData => newData );
```

noisysocks replied:
Is keeping it in `@wordpress/preferences` an option?

talldan replied:
`@wordpress/preferences` is not supposed to have concrete implementation code either, just like `@wordpress/data`. I think this new package is still a better option.

And I think it did need to be moved. The migration code was in the old persistence plugin, but that will become deprecated and unused, we wouldn't want to keep it active just for the purpose of doing a migration.

noisysocks commented on packages/e2e-test-utils/src/preferences.js:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Clears all user meta preferences.
+ */
+export async function resetPreferences() {
```
Did you consider not initialising `@wordpress/database-persistence-layer` in the E2E tests so that only `localStorage` is used in tests? Then this code would be unnecessary.

I suppose it's not ideal because we would then not be testing the "real" application. But arguably we already aren't testing the "real" application. For example we disable CSS animations.

Just a thought to discuss! I don't have a preference.

talldan replied:
I did, but didn't because of the reason you mention, I think it's better to make it as close to a real application as possible.

noisysocks commented on packages/preferences/README.md:
```diff
@@ -12,12 +12,28 @@ npm install @wordpress/preferences --save
+
+### Value
+
+Values can be of any type, but the types supported may be limited by the persistence layer configure. For example of preferences are saved to browser localStorage in JSON format, only JSON serializable types should be used.
```
```suggestion
Values can be of any type, but the types supported may be limited by the persistence layer configure. For example if preferences are saved to browser localStorage in JSON format, only JSON serializable types should be used.
```

noisysocks commented on packages/preferences/CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
 
+### Breaking change
+
+-   The preferences package is no longer compatible with the `@wordpress/data` persistence plugin. Please use the new `setPersistenceLayer` API. ([#39795](https://github.com/WordPress/gutenberg/pull/39795))
```
Is anybody using this package? I doubt it. But if so it might be good to ease the transition here by providing a default `localStorage` persistence layer.

```js
import { localStoragePersistenceLayer } from '@wordpress/preferences';
wp.data.dispatch( 'core/preferences' ).setPersistenceLayer( localStoragePersistenceLayer );
```

talldan replied:
There was one, I removed it in this commit - https://github.com/WordPress/gutenberg/pull/39795/commits/c171bb53b061b5dd07b95c71c1d51176b880a9e4

The main reason being that shipping unused code is something the project tends to avoid. It's also not straightforward to have a default, as the data is fetched asynchronously in an action. I did a lot of headscratching, but couldn't figure out how to make that work.

noisysocks commented on packages/preferences/src/store/index.js:
```diff
@@ -25,14 +22,6 @@ export const store = createReduxStore( STORE_NAME, {
-	selectors,
-	persist: [ 'preferences' ],
-} );
+register( store );
```
🎉

noisysocks commented on packages/preferences/src/store/actions.js:
```diff
@@ -47,3 +47,36 @@ export function setDefaults( scope, defaults ) {
+ * @return {Object} Action object.
+ */
+export async function setPersistenceLayer( persistenceLayer ) {
+	const persistedData = await persistenceLayer.get();
```
Did you consider an approach where `persistenceLayer.get` is called whenever the `get` selector is called? I'm curious what your thoughts were.

The limitation with this approach is that if a preference changes during the lifecycle of the application (e.g. the user adjusts a setting in a different tab) then we have no way of knowing that.

It's probably fine, just thinking aloud.

talldan replied:
I did think about that. I don't think it's worth it, at least for this first iteration.

For the UI to be responsive, `preferences.get` needs to be synchronous. 

The store could be updated with server data in the background, but then there would need to be some handling of data consistency. If a user performs an action that updates a preference while a call to `persistenceLayer.get()` is in flight, the user's optimistic action should overwrite the in flight API request. Which is not actually easy with data that's in a deep structure and can contain arrays.

Ideally I think it would be a push system (server updates dispatch to the store just like user actions) instead of a pull system, and then this would be easier. Because of that it might be something to try when exploring collaborative editing, as there will likely be some need for that kind of server infrastructure.

noisysocks reviewed (commented):
I can sense Andrew Duthie's satisfaction at this! 😂 

Looks really good! I tried to focus mostly on high level considerations. Will give it a spin, review the tests, and proofread the documentation in my next review round.

noisysocks commented:
> If anyone is brave enough, we can review this all in one go and try to merge the one PR.

Most of the diff is moving the legacy preference migrations around so I'm happy to stick with one PR. (Also I like to suffer.)

talldan committed:
Use an IIFE for the inline script and and convert to old school JS

talldan committed:
Use edit context

talldan committed:
Use ISO date for preferences and rename to modified (not a timestamp any more)

talldan committed:
Add clarifying comment about blog prefix

talldan committed:
Fix typo

Co-authored-by: Robert Anderson <robert@noisysocks.com>

talldan committed:
Use only one underscore

talldan committed:
Make asnync debounce more traditional

talldan committed:
Write tests for debounce async function

talldan committed:
Fix timestamps in persistence layer tests

talldan committed:
Use fake timers

talldan commented:
Getting pretty close to addressing all the feedback. Tasks that I think are outstanding:
- Improve the inline scripts / PHP code.
- Rename the package
- Address any further feedback

talldan committed:
Split the PHP hooks in two

talldan committed:
Minimize code in inline script

talldan committed:
Rename database-persistence-layer to preferences-persistence

talldan commented:
Hopefully everything has been addressed now.

In the last commit I renamed the package to 'preferences-persistence'.

noisysocks commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -23,7 +23,15 @@ function gutenberg_configure_persisted_preferences() {
+							'type'        => 'string',
+							'format'      => 'date-time',
+							'context'     => array( 'edit' ),
+							'readonly'    => true,
```
Is `readonly` right here? We're modifying it via REST API put requests, no? Possible that I misunderstand what `readonly` means in this context 😀

talldan replied:
I managed to miss this comment and I've already clicked merge. I'll make a separate PR.

noisysocks replied:
All good—easier to iterate than lug this big PR around.

Let me know where you landed with https://github.com/WordPress/gutenberg/pull/39795#discussion_r854760887 too which is a comment that I forgot about 😀

talldan replied:
Github makes it too easy to miss these things.

You do get an uncaught promise error if the user is offline, or the REST API errors. It doesn't stop the editor from working or anything, so doesn't seem like a critical issue.

Local storage is there as a back up, so everything still works.

I'll work on a PR to see what the options are for handling this better.

noisysocks commented:
> There are also some technical advantages that this will unlock:
> 
> * deprecating the old persistence system in `@wordpress/data`
> * deprecating `registerStore`in `@wordpress/data` in favor of `register`

To clarify: these changes are for follow-ups, right? Making sure it's correct that I don't see them in this PR.

noisysocks reviewed (approved):
Tested this by creating a few different users, switching between them (including using different browsers), and setting different preference combinations. Preferences stay with the user as expected ✨ 

The code looks good and is easy to comprehend! Thanks for adding tests for the new APIs.

Nice work. Let's get it in!

talldan commented:
> To clarify: these changes are for follow-ups, right? Making sure it's correct that I don't see them in this PR.

Yep, exactly.

Thanks for all the reviews for this mini project, really appreciate them.

Mamaduka commented:
Fantastic work, @talldan 🥇

youknowriad commented:
Thanks for pushing this into the finish line @noisysocks and @talldan

youknowriad commented:
This PR improve the "block selection" (focus) performance by half 😮 This is awesome but I guess it was not intentional so it would be good to try to understand why it's the case https://codehealth.vercel.app

talldan commented:
> This PR improve the "block selection" (focus) performance by half 😮 This is awesome but I guess it was not intentional so it would be good to try to understand why it's the case https://codehealth.vercel.app/

Wow. I don't know why. 

It seems weird that it's block selection. There's not really a flow that has any relation to preferences.

Mamaduka commented:
Is it possible to disable persisting preferences in DB for some cases?

I noticed that an API request is triggered every time [switch between](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/editor-initialization/listener-hooks.js#L41-L45) a Document and a block.

https://user-images.githubusercontent.com/240569/165691256-7f689859-9379-4094-a238-e91a4ce64ca5.mp4

talldan commented:
@Mamaduka It's a valid concern. At the moment there isn't a way to handle this. I think it would require some extra configuration for particular preferences.

A possibility is that some preferences don't trigger a request. They could save to local storage, and maybe they do persist to user meta too, but they could only update when another preference is saved AND/OR only update after a particular period of time AND/OR only update on browser unload (it's difficult to ensure this works though).

Mamaduka commented:
I was thinking about a similar method. For example, we could use post autosave to trigger preference syncing with user meta.

youknowriad commented:
>I noticed that an API request is triggered every time [switch between](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/editor-initialization/listener-hooks.js#L41-L45) a Document and a block.

Interesting, I wonder if that's the reason for the performance gain (now that we don't use localstorage which might have been synchronous)

Anyway, it does feel like switching between sidebars shouldn't be a preference at all. Whether the sidebar is closed or open should be, so there's a small refactoring to be done here to improve things.

gziolo commented:
I opened https://github.com/WordPress/gutenberg/pull/40715 with minor edits to the package configuration.

noisysocks commented:
> Anyway, it does feel like switching between sidebars shouldn't be a preference at all. Whether the sidebar is closed or open should be, so there's a small refactoring to be done here to improve things.

Yeah agree this doesn't need to be persisted across sessions.

talldan commented:
> Anyway, it does feel like switching between sidebars shouldn't be a preference at all. Whether the sidebar is closed or open should be, so there's a small refactoring to be done here to improve things.

I've made a PR that does this - #40923

talldan commented:
There may be a few other things we also want to refactor in this way. Expanded/collapsed panels in the post editor's document settings is something we could consider removing IMO.

I also have a PR to make the block editor package's `insertUsage` use the preferences package - https://github.com/WordPress/gutenberg/pull/39632.

It would result in a REST API request every time a block is inserted. This might be a more challenging one to solve.

Mamaduka commented:
> There may be a few other things we also want to refactor in this way. Expanded/collapsed panels in the post editor's document settings is something we could consider removing IMO.

I also noticed that.

Maybe API can support a list of preferences that don't need to be persisted immediately.

mcsf commented on lib/compat/wordpress-6.1/persisted-preferences.php:
```diff
@@ -0,0 +1,106 @@
+		sprintf(
+			'( function() {
+				var serverData = %s;
+				var userId = "%s";
```
Minor and belated, but it seems safer to encode `userId` formally, in the same way as `serverData`, even if we know that the result

```php
sprintf( 'userId = %s', wp_json_encode( $user_id ) );
```

would be identifical to the current form

```php
sprintf( 'userId = "%s"', $user_id );
```

The former doesn't rely on our own mental type checking, and redundancy helps with safety — especially around code interpolation. :)

swissspidy commented:
@talldan Curious, why doesn’t this use the existing user settings API in WordPress (`get_all_user_settings()` and friends)? These are already stored in user meta.

talldan commented:
@swissspidy There was a past attempt to use user settings, which discusses some of the reasons - https://github.com/WordPress/gutenberg/pull/15800.

therealgilles commented:
I just filed this [bug report](https://github.com/WordPress/gutenberg/issues/60325). I don't think this code should be making API calls to save user preferences if the user is not logged in.
